core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/NetworkTopologyReplicationStrategy.java

```diff
@@ -139,11 +147,20 @@
         }
       }
 
-      result.putAll(ring.get(i), replicas);
+      Set<Node> canonicalReplicaSet =
+          canonicalReplicaSets.computeIfAbsent(replicas, ImmutableSet::copyOf);
+      result.put(ring.get(i), canonicalReplicaSet);
     }
     return result.build();
   }
 
+  private static void addIfAbsent(List<Node> nodes, Node node) {
```

Comment:
There is a small `CollectionsUtils` class, maybe you could move this utility method there.

Reply:
I ended up extracting a `CanonicalNodeSetBuilder` that is reused across the two strategies, it's a nice way to avoid duplication.